rfcs/20200424-sparse-domain-isolation.md

```diff
@@ -0,0 +1,649 @@
+                         name="wide-sparse-weights")
+                         # z should be a trainable and variable-size instance.
+# forward graph
+
+pred = tf.XXX(...)
+
+# backward graph
+loss = tf.XXX(pred, label, ...)
+opt = tf.train.AdamOptimizer(learning_rate=0.1)
+update_op = opt.minimize(loss)
+```
+
+## Sparse Domain Isolation
+
+### Overview of Design
+![Overview Flow Chart](20200424-sparse-domain-isolation/overview-flow-chart.png)
+
+### Design Considerations
+
+![Architecture of SDI](20200424-sparse-domain-isolation/architecture.png)
+*   Minimize changing on Tensorflow core and try to use native components.
+
+![Expression Weights Hierarchically](20200424-sparse-domain-isolation/expression-weights-hierarchically.png)
+*   The `tf.Tensor` is still suitable for holding iteration weights, because only a small part of weights will be updated in each iterations.
+
+#### Hash Table
+
+Refer to `MutableHashTable` implemented in `tf.lookup`. The reason is as follows:
+
+*   The `MutableHashTable` meets the basic requirements of sparse weights.
+*   The `tf.lookup` is well designed and easy to maintain.
+*   The save/restore from/to checkpoint is already supported.
+*   Compatible with all kind of **distributed strategy**(including parameters server.)
+
+
+### Detailed Design and Implementation
+
+The trainable warp class `resource_variable_ops.TrainableWrap` inherted from `ResourceVariable` will be introduced, and help us reuse all optimizers.
+
+![detail of SDI](20200424-sparse-domain-isolation/optimizers-reuse-scheme.png)
+
+### APIs Overview
+
+* Name Space(**To Be Discussed**)
+    *  USER API: `tf.dynamic_embedding`
+    *  Tensorflow Core: `from tensorflow.xx.xx import dynamic_embedding_ops`  
+    
+    
+* Creating Weights
+    * `tf.dynamic_embedding.Variable`
+    * `tf.dynamic_embedding.get_variable`
+
+* Trainable Wrap
+    * `tf.resource_variable_ops.TrainableWrap`
+    * `tf.dynamic_embedding.embedding_lookup`
+    * `tf.dynamic_embedding.embedding_lookup_sparse`
+
+#### Name Space
+
+All of the APIs are implemented in `tf.dynamic_embedding` package.
+Maybe the `tf.nn` or `tf.keras` would be better choices.
+
+```python
+import tensorflow as tf # core version
+import tensorflow_recsys as tfrs # SIG version
+
+weights = tf.dynamic_embedding.get_variable(*args, **kwargs)
+```
+
+#### Creating Sparse Weights
+
+We design the `dynamic_embedding.Variable` based on the `tf.lookup.MutableHashTable` and make it support distributed scenario, there is no need to develop any new operators for the hash table. 
+The `dynamic_embedding.Variable` backed up by a group of hashtables is responsible for actually holding the memory resource of sparse weights.
+
+* `tf.dynamic_embedding.Variable`
+    * `export`
+    * `remove`
+    * `upsert` (update if exist else insert) 
+    * `size`
+* `tf.dynamic_embedding.get_variable`
+
+```python
+@tf_export("dynamic_embedding.Variable")
+class Variable(object):
```

Comment:
We probably can use a different name instead of Variable?

Reply:
> We probably can use a different name instead of Variable?

Yeah, agree with you.